### PR TITLE
Releng/migrate to crt

### PR DESCRIPTION
**Pull request by emilymianeil**
These are the changes necessary in order to migrate `waypoint` from CircleCI to CRT. As such, `waypoint` will be able to release from this repository as opposed to using the `waypoint-releases` repository. We need @hashicorp/waypoint to confirm the questions commented in this PR before signing off. 

The following is a summary of changes made:
-  Building all binaries for CRT pipeline
- Create custom Make targets for CRT
- Added `ci.hcl` to define the CRT pipeline

# TODO Items
- [X] @-hashicorp/waypoint please let us know the list of @hashicorp/release-approvers-waypoint-oss  you would like to approve releases
- [ ] RelEng - remove current branch reference in `ci.hcl` and `build.yaml`

### CODE REVIEW

File: .github/workflows/build-waypoint-ent.yml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+          product_version: ${{ inputs.waypoint-version }}
+          go_version: ${{ inputs.go-version }}
+          os: ${{ inputs.goos }}
+          arch: ${{ inputs.goarch }}
+          reproducible: report
+          instructions: |
+            make bin/crt-${{ inputs.package-name }}
+        # FIXME: are we still only publishing linux amd64 packages?
+      - name: Linux packaging
+        if: ${{ (inputs.goos == 'linux') && (inputs.goarch == 'amd64') && (inputs.package-name == 'waypoint') }}
```
Comment (emilymianeil) [inline .github/workflows/build-waypoint-ent.yml:89]: @waypoint - Please confirm that the only arch supported for linux packages will be `linux amd64`

File: .github/workflows/build-waypoint-oss.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        with:
+          product_name: ${{ inputs.package-name }}
+          product_version: ${{ inputs.waypoint-version }}
+          go_version: ${{ inputs.go-version }}
+          os: ${{ inputs.goos }}
+          arch: ${{ inputs.goarch }}
+          reproducible: report
+          instructions: |
+            make bin/crt-${{ inputs.package-name }}
+        # FIXME: are we still only publishing linux amd64 packages?
```
Comment (emilymianeil) [inline .github/workflows/build-waypoint-oss.yml:86]: @waypoint - Please confirm that the only arch we are supporting for linux packaging is `amd64`?

File: .github/workflows/build-waypoint-ent.yml (hunk 1 above)
Comment (briancain) [reply .github/workflows/build-waypoint-ent.yml:89]: We support both `amd64` and `arm` packages on linux.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+      - name: Save assets to cache
+        uses: actions/cache/save@v3
+        with:
+          path: ui/dist/assets
+          key: ui_assets_${{ github.sha }}
+
+  build-assets:
+    # FIXME: can releases be tagged as per CRT standards (ie at the end of the pipeline)?
+    runs-on: ubuntu-latest
+    name: Build Assets
```
Comment (emilymianeil) [inline .github/workflows/build.yml:111]: The current `waypoint-releases` repo seems to tag the commit at the beginning of the pipeline as seen [here](https://github.com/hashicorp/waypoint-releases/blob/main/.circleci/config.yml#L368). However, CRT typically tags releases on `promote-to-production` - is @waypoint ok with RelEng tagging releases at the end of pipeline?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+      - get-product-version
+      - get-go-version
+      - build-assets
+      - build-ui
+    strategy:
+      matrix:
+        pkg-name: [ "waypoint", "waypoint-entrypoint" ]
+        goos: [ linux ]
+        goarch: [ "386", "amd64", "arm", "arm64" ]
+        # FIXME: can we confirm which os/arch combos we require for this pipeline?
```
Comment (emilymianeil) [inline .github/workflows/build.yml:149]: @waypoint - Please confirm which os/arch combinations are necessary for this pipeline. As of now we have:

1. Linux/386 (+entrypoint)
2. Linux/amd64 (+entrypoint)
3. Linux/arm (+entrypoint)
4. Linux/arm64 (+entrypoint)
5. Darwin/amd64 (+entrypoint)
6. Darwin/arm64 (+entrypoint)
7. Windows/386
8. Windows/amd64

File: .github/workflows/build-waypoint-oss.yml (hunk 2 above)
Comment (emilymianeil) [reply .github/workflows/build-waypoint-oss.yml:86]: Update to `amd64` and `arm` for linux packages

File: .github/workflows/build.yml (hunk 4 above)
Comment (emilymianeil) [reply .github/workflows/build.yml:149]: Correct: https://hashicorp.slack.com/archives/C013QT1KG9W/p1679517514985589

File: Dockerfile [hunk 5]
```diff
@@ -81,3 +81,89 @@ ENV HOME /home/waypoint
...
+# NAME and PRODUCT_VERSION are the name of the software in releases.hashicorp.com
+# and the version to download. Example: NAME=boundary PRODUCT_VERSION=1.2.3.
+ARG NAME=waypoint
+ARG PRODUCT_VERSION
+# TARGETARCH and TARGETOS are set automatically when --platform is provided.
+ARG TARGETOS
+ARG TARGETARCH
+
+LABEL name="Waypoint" \
+      maintainer="HashiCorp Waypoiny Team <waypoint@hashicorp.com>" \
```
Comment (catsby) [inline Dockerfile:101]:
```suggestion
      maintainer="HashiCorp Waypoint Team <waypoint@hashicorp.com>" \
```

File: Dockerfile [hunk 6]
```diff
@@ -81,3 +81,89 @@ ENV HOME /home/waypoint
...
```
Comment (catsby) [inline Dockerfile:151]:
```suggestion
      maintainer="HashiCorp Waypoint Team <waypoint@hashicorp.com>" \
```

File: .changelog/4557.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+core: Migrate to CRT.
```
Comment (catsby) [inline .changelog/4557.txt:2]: Have other products included this in their CHANGELOG? It doesn't seem user-facing so personally I would not include it, as our end-users have no real concept of what CRT is and how it applies to them (it doesn't)

**Review by catsby — COMMENTED** [review]
Some initial feedback as I continue to review

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+      - name: Prepare zipped binaries to build Docker image
+        id: prepare-zips
+        run: |
+          DIR=$(mktemp -d)
+          echo "working-dir=$DIR" >> $GITHUB_OUTPUT
+          cp -v Dockerfile $DIR/Dockerfile
+          unzip -j archive/waypoint_${{env.version}}_linux_${{ matrix.arch }}.zip -d $DIR
+          unzip -j archive/waypoint-entrypoint_${{env.version}}_linux_${{ matrix.arch }}.zip -d $DIR
+      - name: Build Docker image
+        uses: hashicorp/actions-docker-build@d960286e523e71dec7d9571a52a5a0c6b9ffdee7
```
Comment (dekimsey) [inline .github/workflows/build.yml:207]: `v1.5.0` has this change, we'll update `v1` tomorrow

**Review by dekimsey — APPROVED** [review]

File: .changelog/4557.txt (hunk 7 above)
Comment (emilymianeil) [reply .changelog/4557.txt:2]: They've not, the Changelog-checker workflow was failing so I added this but I can definitely remove this

File: .github/workflows/build-waypoint-oss.yml [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+          name: ${{ inputs.package-name }}
+          description: ${{ inputs.package-name }}
+          arch: ${{ inputs.goarch }}
+          version: ${{ inputs.waypoint-version }}+${{ inputs.waypoint-metadata }}
+          maintainer: "HashiCorp"
+          homepage: "https://github.com/hashicorp/waypoint"
+          license: "MPL-2.0"
+          binary: "dist/${{ inputs.package-name }}"
+          deb_depends: "git"
+          rpm_depends: "git"
```
Comment (dekimsey) [inline .github/workflows/build-waypoint-oss.yml:106]: Is this an actual runtime dependency of waypoint?

**Review by dekimsey — APPROVED** [review]

File: .github/workflows/build-waypoint-oss.yml (hunk 9 above)
Comment (catsby) [reply .github/workflows/build-waypoint-oss.yml:106]: We do use submodules regarding 3rd party protocol buffers and run a few `git` commands in other makefile steps, would those reasons explain this need?

**Review by catsby — APPROVED** [review]
I worked with @dekimsey and was able to push new builds and verify them, I _think_ we're good here at least as far as the Waypoint team is concerned

File: .github/workflows/build-waypoint-oss.yml (hunk 9 above)
Comment (dekimsey) [reply .github/workflows/build-waypoint-oss.yml:106]: 100%

File: version/VERSION
Comment (dekimsey) [inline version/VERSION:1]: How have we handled this addition in other products? I mownedring what happens during backport, if we make a note to correct it or if we should pull this out and merge this change into main before this PR.

**Review by dekimsey — APPROVED** [review]

**Review by dekimsey — APPROVED** [review]
LGTM

File: version/VERSION
Comment (dekimsey) [reply version/VERSION:1]: We'll fix on backport

File: .github/workflows/build-waypoint-oss.yml (hunk 9 above)
Comment (catsby) [reply .github/workflows/build-waypoint-oss.yml:106]: > Do end-users need git installed to use waypoint?

No

**Pull request merged**